src/Microsoft.OData.Core/ODataWriterCore.cs

```diff
@@ -85,7 +85,7 @@ internal abstract class ODataWriterCore : ODataWriter, IODataOutputInStreamError
             this.listener = listener;
 
             this.scopeStack.Push(new Scope(WriterState.Start, /*item*/null, navigationSource, resourceType, /*skipWriting*/false, outputContext.MessageWriterSettings.SelectedProperties, odataUri));
-            this.CurrentScope.DerivedTypeConstraints = this.outputContext.Model.GetDerivedTypeConstraints(navigationSource);
+            this.CurrentScope.DerivedTypeConstraints = this.outputContext.Model.GetDerivedTypeConstraints(navigationSource)?.ToList();

```

Comment:
Making sure that `GetDerivedTypeConstraints` returns an empty collection rather than `null` would probably save us some trouble but from the look of things it appears like that would be a bigger change.